### PR TITLE
Macpie/sc server rework

### PR DESCRIPTION
**Pull request by macpie**

### CODE REVIEW

File: src/state_channel/blockchain_state_channels_server.erl [hunk 1]
```diff
@@ -273,13 +264,15 @@ handle_info({blockchain_event, {add_block, BlockHash, _Syncing, _Ledger}}, #stat
     {noreply, State};
 handle_info({got_block, Block, _BlockHash, []}, State0) ->
     Height = blockchain_block:height(Block),
+    State1 = State0#state{height=Height},
     lager:debug("no transactions found in ~p", [Height]),
-    State1 = check_state_channel_expiration(Block, State0),
-    {noreply, State1#state{height=Height}};
+    ok = check_state_channel_expiration(Block, State1),
+    {noreply, State0};
```
Comment (vihu) [inline src/state_channel/blockchain_state_channels_server.erl:270]: Is this a typo? `State1` seems to have the newer block height...

File: src/blockchain_utils.erl [hunk 2]
```diff
@@ -187,25 +182,32 @@ pfind(F, ToDos) ->
...
-        [],
-        ToDos
+        Opts
     ),
-    Results = pfind_rcv(Ref, false, erlang:length(ToDos)),
-    [erlang:exit(Pid, done) || Pid <- Workers],
-    Results.
+    receive
+        {Ref, Results} ->
+            Results
```
Comment (michaeldjeffrey) [inline src/blockchain_utils.erl:209]: There's a chance if a child dies before anyone returns an answer that everyone will die, and we'll hang here forever.

File: src/blockchain_utils.erl [hunk 3]
```diff
@@ -813,6 +817,12 @@ pfind_test() ->
     end,
     Args = [[I] || I <- lists:seq(1, 6)],
     ?assertEqual({true, 2}, pfind(F, Args)),
+    receive
+        _ ->
+            ?assert(fasle)
```
Comment (michaeldjeffrey) [inline src/blockchain_utils.erl:822]:
```suggestion
            ?assert(false)
```

File: src/state_channel/blockchain_state_channels_server.erl [hunk 4]
```diff
@@ -195,37 +204,19 @@ handle_cast({gc_state_channels, SCIDs}, #state{state_channels=SCs}=State) ->
...
-    {noreply, State1};
-handle_cast(get_new_active, State0) ->
-    lager:info("get a new active state channel"),
-    State1 = get_new_active(State0),
-    {noreply, State1};
-handle_cast({handle_worker_terminate, SC, Reason}, #state{state_channels=SCs}=State) ->
-    ID = blockchain_state_channel_v1:id(SC),
-    lager:info("~p is terminated ~p", [blockchain_utils:addr2name(ID), Reason]),
-    {noreply, State#state{state_channels=maps:put(ID, SC, SCs)}};
+handle_cast({update_state_channel, UpdatedSC}, #state{state_channels=SCs0}=State) ->
```
Comment (michaeldjeffrey) [inline src/state_channel/blockchain_state_channels_server.erl:208]: the channels worker will terminate itself if `expired` or `overspent`, do those states need to be sent here?

File: src/state_channel/blockchain_state_channels_server.erl [hunk 5]
```diff
@@ -322,6 +332,14 @@ terminate(_Reason, _State) ->
     HandlerPid :: pid()
 ) -> ok | reject.
 handle_offer(Offer, Ledger, HandlerPid) ->
+    handle_offer_(Offer, Ledger, HandlerPid, 2).
+
+handle_offer_(Offer, _Ledger, _HandlerPid, 0) ->
+    HotspotID = blockchain_state_channel_offer_v1:hotspot(Offer),
+    HotspotName = blockchain_utils:addr2name(HotspotID),
+    lager:debug("could not handle offer too may retries for ~p", [HotspotName]),
```
Comment (michaeldjeffrey) [inline src/state_channel/blockchain_state_channels_server.erl:340]:
```suggestion
    lager:debug("could not handle offer too many retries for ~p", [HotspotName]),
```

File: src/state_channel/blockchain_state_channels_worker.erl [hunk 6]
```diff
@@ -223,7 +230,9 @@ offer(
                 "dropping this packet because it will overspend DC ~p, (cost: ~p, total_dcs: ~p)",
                 [DCAmount, NumDCs, TotalDCs]
             ),
-            {stop, {shutdown, overspent}, State0};
+            %% This allow for packets (accepted offer) to come threw 
+            _ = erlang:send_after(1000, self(), ?OVERSPENT),
```
Comment (michaeldjeffrey) [inline src/state_channel/blockchain_state_channels_worker.erl:234]: This offer was just rejected. 
What are the packets that are coming through?

File: src/state_channel/blockchain_state_channels_server.erl [hunk 7]
```diff
@@ -62,25 +67,26 @@ start_link(Args) ->
...
             end,
-            gen_server:call(?SERVER, get_actives, infinity)
+            maps:values(gen_server:call(?SERVER, get_actives))
         )
     ).
 
 -spec get_active_pid(ID :: blockchain_state_channel_v1:id()) -> pid() | undefined.
 get_active_pid(ID) ->
-    gen_server:call(?SERVER, {get_active_pid, ID}, infinity).
+    gen_server:call(?SERVER, {get_active_pid, ID}).
```
[the remark this answers is not in the document]
Comment (macpie) [reply src/state_channel/blockchain_state_channels_server.erl:89]: yes 5s default, this cause some issue that halted SC opening on prod

File: src/blockchain_utils.erl (hunk 2 above)
Comment (macpie) [reply src/blockchain_utils.erl:209]: yes this was the case for the old implementation also,not sure if we can put an appropriate time here tho

Comment (Vagabond) [reply src/blockchain_utils.erl:209]: you can use the `monitor` option to spawn_opt and you'll get back `{pid(), reference()}`

**Review by michaeldjeffrey — APPROVED** [review]

**Pull request merged**